In [87]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt

In [88]:
df = pd.read_csv("/Users/mac/Desktop/projects/zoomcamp ml/Lesson 1/laptops.csv")
df.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

In [89]:
df.describe()

RAM      Storage       Screen  Final Price
count  2160.000000  2160.000000  2156.000000  2160.000000
mean     15.413889   596.294444    15.168112  1312.638509
std       9.867815   361.220506     1.203329   911.475417
min       4.000000     0.000000    10.100000   201.050000
25%       8.000000   256.000000    14.000000   661.082500
50%      16.000000   512.000000    15.600000  1031.945000
75%      16.000000  1000.000000    15.600000  1708.970000
max     128.000000  4000.000000    18.000000  7150.470000

## Q2. Records count
###  How many records are in the dataset?

### Answer: 2160

In [90]:
df.shape[0]

2160

## Q3. Laptop brands
### How many laptop brands are presented in the dataset?

### Answer: 27

In [91]:
df['Brand'].nunique()

27

## Q4. Missing values
### How many columns in the dataset have missing values?

### Answer: 3

In [92]:
df.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

## Q5. Maximum final price
### What's the maximum final price of Dell notebooks in the dataset?

### Answer: 3936.00

In [93]:
df.groupby('Brand')['Final Price'].max()

Brand
Acer                3691.00
Alurin               869.00
Apple               3849.00
Asus                5758.14
Deep Gaming         1639.01
Dell                3936.00
Denver               329.95
Dynabook Toshiba    1805.01
Gigabyte            3799.00
HP                  5368.77
Innjoo               431.38
Jetwing              469.27
LG                  2399.00
Lenovo              5018.14
MSI                 7150.47
Medion              3799.00
Microsoft           3747.91
Millenium           2312.71
PcCom               1949.90
Primux               599.41
Prixton              329.95
Razer               4999.01
Realme               999.00
Samsung             3699.01
Thomson              436.56
Toshiba              799.00
Vant                1217.01
Name: Final Price, dtype: float64

## Q6. Median value of Screen
 #### 1  Find the median value of Screen column in the dataset.
 #### 2  Next, calculate the most frequent value of the same Screen column.
 #### 3   Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
 #### 4 Now, calculate the median value of Screen once again.



### Answer: No it didn't change


In [94]:
df['Screen'].median()

15.6

In [95]:
df['Screen'].mode()

0    15.6
Name: Screen, dtype: float64

In [96]:
df['Screen'].fillna(df['Screen'].mode()[0], inplace=True)

In [97]:
df['Screen'].median()

15.6

## Q7. Sum of weights
   ### Select all the "Innjoo" laptops from the dataset.
   ### Select only columns RAM, Storage, Screen.
   ### Get the underlying NumPy array. Let's call it X.
   ### Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
   ### Compute the inverse of XTX.
   ### Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
   ### Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
   ### What's the sum of all the elements of the result?


## Answer: 91.30

In [98]:
df.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

In [108]:
injoo_brand = df[df['Brand'] == "Innjoo"]
injoo_brand = injoo_brand[['RAM', "Storage", 'Screen']]
X = injoo_brand.values
XTX = np.dot(X.T, X)
XTX_inv = np.linalg.inv(XTX)
y = np.array([1100, 1300, 800, 900, 1000, 1100]).reshape(-1, 1) 
w = np.dot(np.dot(XTX_inv, X.T), y)
result = np.sum(w)
result



91.2998806299555

In [106]:
w

array([[45.58076606],
       [ 0.42783519],
       [45.29127938]])

In [102]:
XTX.shape

(3, 3)

In [103]:
XTX_inv.shape

(3, 3)